domains/certificates/Query.API/API/Query.API/Projections/CertificatesByOwnerProjection.cs

```diff
@@ -1,119 +1,186 @@
     {
-        if (view.Owner == @event.Source)
-        {
-            view.Certificates.Remove(@event.CertificateId);
-        }
-        else
+        var duplicatedStream = streams
+            .SelectMany(x => x.Events)
+            .DistinctBy(x => x.StreamId);
+
+        if (duplicatedStream.Count() != 1)
+            throw new ProjectionException(Guid.Empty, "Only one certificate is allowed in the stream");
+
+        var events = streams
+            .SelectMany(x => x.Events)
+            .OrderBy(s => s.Sequence)
+            .Select(s => s.Data);
+
+        foreach (var @event in events)
         {
-            view.Certificates[@event.CertificateId] = new CertificateView
+            switch (@event)
             {
-                CertificateId = @event.CertificateId,
-                DateFrom = @event.Period.DateFrom,
-                DateTo = @event.Period.DateTo,
-                Quantity = @event.ShieldedQuantity.Value,
-                GSRN = @event.ShieldedGSRN.Value,
-                GridArea = @event.GridArea,
-                TechCode = @event.Technology.TechCode,
-                FuelCode = @event.Technology.FuelCode,
-                Status = CertificateStatus.Issued // Transfer is only allowed if certificate is issued
-            };
+                case ProductionCertificateCreated productionCertificateCreated:
+                    CreateCertificatesByOwnerView(operations,
+                        productionCertificateCreated);
+                    break;
+                case ProductionCertificateIssued productionCertificateIssued:
+                    ApplyProductionCertificateIssued(operations,
+                        productionCertificateIssued);
+                    break;
+                case ProductionCertificateRejected productionCertificateRejected:
+                    ApplyProductionCertificateRejected(
+                        operations, productionCertificateRejected);
+                    break;
+                case ProductionCertificateTransferred productionCertificateTransferred:
+                    ApplyProductionCertificateTransferred(operations, productionCertificateTransferred);
+                    break;
+                default:
+                    throw new InvalidOperationException();
+            }
         }
+
+        operations.Store(views.ToArray());
+    }
+
+    private static void ApplyProductionCertificateRejected(IDocumentOperations operations,
+        ProductionCertificateRejected productionCertificateRejected)
+    {
+        var view = views
+                       .FirstOrDefault(it =>
+                           it.Certificates.ContainsKey(productionCertificateRejected.CertificateId)) ??
+                   GetCertificatesByOwnerView(operations, productionCertificateRejected.CertificateId);
+
+        view.Certificates[productionCertificateRejected.CertificateId].Status = CertificateStatus.Rejected;
+        views.Add(view);
```

Comment:
If the item is already in `views`, that item will be added again to `views`. I think it would be safer to *not* have the same item multiple times in `views`